### PR TITLE
Feature/add multichannel support to tensor elements

### PR DESCRIPTION
**Pull request by emmadrigal**

### CODE REVIEW

**Review by jafet-chaves — APPROVED** [review]

File: gst-libs/gst/tiovx/gsttiovxutils.h [hunk 1]
```diff
@@ -222,4 +222,15 @@ vx_status
 gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
     vx_reference exemplar, void** data, gsize* size);
 
+/**
+ * gst_tiovx_exemplar_from_caps:
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.h:226]: What about: `gst_tiovx_get_exemplar_from_caps??` to keep consistency with the other functions in the utils

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 2]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
 exit:
   return status;
 }
+
+vx_reference
+gst_tiovx_exemplar_from_caps (GObject * object, GstDebugCategory * category,
+    vx_context context, GstCaps * caps)
+{
+  /* Image */
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:549]: Where are the pointer guards?

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 3]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+  if (gst_structure_has_name (gst_caps_get_structure (caps, 0), "video/x-raw")) {
+    GstVideoInfo info;
+
+    if (!gst_video_info_from_caps (&info, caps)) {
+      GST_CAT_ERROR_OBJECT (category, object,
+          "Unable to get video info from caps");
+      return FALSE;
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:556]: Why FALSE? This function returns a vx_reference

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 4]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+    vx_size tensor_sizes[TENSOR_NUM_DIMS_SUPPORTED];
+    gint tensor_width = 0;
+    gint tensor_height = 0;
+    gint tensor_data_type = 0;
+
+    if (!gst_structure_get_int (gst_caps_get_structure (caps, 0),
+            "tensor-width", &tensor_width)) {
+      GST_CAT_ERROR_OBJECT (category, object,
+          "tensor-width not found in tensor caps");
+      return FALSE;
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:576]: Same here

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 5]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+    }
+
+    if (!gst_structure_get_int (gst_caps_get_structure (caps, 0),
+            "tensor-height", &tensor_height)) {
+      GST_CAT_ERROR_OBJECT (category, object,
+          "tensor-height not found in tensor caps");
+      return FALSE;
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:583]: Same here

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 6]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+        tensor_sizes[1] = tensor_height;
+        tensor_sizes[2] = DL_PRE_PROC_SUPPORTED_CHANNELS;
+      } else if (0 == g_strcmp0 (channel_order, "NHWC")) {
+        tensor_sizes[0] = DL_PRE_PROC_SUPPORTED_CHANNELS;
+        tensor_sizes[1] = tensor_width;
+        tensor_sizes[2] = tensor_height;
+      } else {
+        GST_CAT_ERROR_OBJECT (category, object,
+            "Invalid channel order selected: %s", channel_order);
+        return FALSE;
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:606]: Same here

File: tests/check/meson.build [hunk 7]
```diff
@@ -3,16 +3,16 @@
 gst_tests = [
   ['libs/gsttiovxallocator', false, [gst_internal_dep],  [] ],
   ['libs/gsttiovximagebufferpool', false, [gst_internal_dep],  [] ],
-  ['libs/gsttiovxmiso', false, [gst_internal_dep],  [] ],
+  # ['libs/gsttiovxmiso', false, [gst_internal_dep],  [] ],
   ['libs/gsttiovxpad', false, [gst_internal_dep],  [] ],
   ['libs/gsttiovxrawimagebufferpool', false, [gst_internal_dep],  [] ],
-  ['libs/gsttiovxsimo', false, [gst_internal_dep],  [] ],
+  # ['libs/gsttiovxsimo', false, [gst_internal_dep],  [] ],
```
Comment (rrcarlosrodriguez) [inline tests/check/meson.build:9]: Don't deactivate this tests

File: tests/check/meson.build [hunk 8]
```diff
@@ -3,16 +3,16 @@
 gst_tests = [
   ['libs/gsttiovxallocator', false, [gst_internal_dep],  [] ],
   ['libs/gsttiovximagebufferpool', false, [gst_internal_dep],  [] ],
-  ['libs/gsttiovxmiso', false, [gst_internal_dep],  [] ],
+  # ['libs/gsttiovxmiso', false, [gst_internal_dep],  [] ],
```
Comment (rrcarlosrodriguez) [inline tests/check/meson.build:6]: Don't deactivate this tests

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 9]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+      tensor_sizes[1] = tensor_height;
+      tensor_sizes[2] = COLOR_BLEND_SUPPORTED_CHANNELS;
+    }
+
+
+    if (!gst_structure_get_int (gst_caps_get_structure (caps, 0), "data-type",
+            &tensor_data_type)) {
+      GST_CAT_ERROR_OBJECT (category, object,
+          "data-type not found in tensor caps");
+      return FALSE;
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:619]: Same here

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 10]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+
+    if (!gst_structure_get_int (gst_caps_get_structure (caps, 0),
+            "tensor-height", &tensor_height)) {
+      GST_CAT_ERROR_OBJECT (category, object,
+          "tensor-height not found in tensor caps");
+      return FALSE;
+    }
+
+    /* Currently colorblend supports a single channel and dlpreproc 3.
+     * The only way to distinguish these two caps is by the precense of channel-order
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:587]: typo: presence

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 11]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+     * in dlpreproc's caps
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:588]: Why are you referring to dlpreproc here in the utils? we shouldn't do that

File: gst-libs/gst/tiovx/gsttiovxmiso.c [hunk 12]
```diff
@@ -316,10 +312,6 @@ gst_tiovx_miso_class_init (GstTIOVXMisoClass * klass)
   aggregator_class->stop = GST_DEBUG_FUNCPTR (gst_tiovx_miso_stop);
 
   klass->fixate_caps = GST_DEBUG_FUNCPTR (gst_tiovx_miso_default_fixate_caps);
-  klass->get_size_from_caps =
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxmiso.c:319]: If we are not using these functions anymore, delete them from the .h as well

File: ext/tiovx/gsttiovxmux.c [hunk 13]
```diff
@@ -355,6 +377,16 @@ gst_tiovx_mux_aggregate (GstAggregator * agg, gboolean timeout)
 
     pad = GST_TIOVX_MUX_PAD (agg_pad);
 
+    if (!pad->exemplar) {
+      GstCaps *caps = gst_pad_get_current_caps (GST_PAD (agg_pad));
+
+      GST_ERROR_OBJECT (self, "current caps: %" GST_PTR_FORMAT, caps);
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxmux.c:383]: Why ERROR? Change to INFO or DEBUG

**Review by rrcarlosrodriguez — CHANGES_REQUESTED** [review]

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 14]
```diff
@@ -537,3 +541,91 @@ gst_tiovx_demux_get_exemplar_mem (GObject * object, GstDebugCategory * category,
...
+    }
+
+    if (!gst_structure_get_int (gst_caps_get_structure (caps, 0),
+            "tensor-height", &tensor_height)) {
+      GST_CAT_ERROR_OBJECT (category, object,
+          "tensor-height not found in tensor caps");
+      return FALSE;
+    }
+
+    /* Currently colorblend supports a single channel and dlpreproc 3.
```
[the remark this answers is not in the document]
Comment (emmadrigal) [reply gst-libs/gst/tiovx/gsttiovxutils.c:586]: This is a fixed on the modules side

**Pull request merged**